src/Dialect/ONNX/Transforms/Decompose.cpp

```diff
@@ -668,37 +669,548 @@ bool shouldDecomposeConvTransposeOp(Value convTransposeResult) {
+  auto padsShape = (padsShapeAttr) ? getVectorFromArrayAttr(padsShapeAttr)
+                                   : SmallVector<int64_t, 4>({0, 0, 0, 0});
+  auto stridesShape = (stridesShapeAttr)
+                          ? getVectorFromArrayAttr(stridesShapeAttr)
+                          : SmallVector<int64_t, 4>({1, 1});
+  ONNXConvTransposeOpShapeHelper convTransposeShapeHelper(
+      op.getOperation(), {});
+  convTransposeShapeHelper.computeShapeAndAssertOnFailure();
+
+  int outputRank = convTransposeShapeHelper.getOutputDims().size();
+  SmallVector<int64_t, 4> outputShape;
+  for (int i = 0; i < outputRank; ++i) {
+    int64_t d = convTransposeShapeHelper.getOutputDims()[i].isLiteral()
+                    ? convTransposeShapeHelper.getOutputDims()[i].getLiteral()
+                    : ShapedType::kDynamic;
+    outputShape.emplace_back(d);
+  }
+  assert(stridesShape.size() == 2);
+  if (stridesShape[0] != stridesShape[1])
+    return false;
+  if (stridesShape[0] > 3)
+    return false;
+  bool hDivisisbleByStride =
+      (outputShape[outputShape.size() - 2] % stridesShape[0] == 0);
+  bool wDivisisbleByStride =
+      (outputShape[outputShape.size() - 1] % stridesShape[0] == 0);
+  if (!hDivisisbleByStride || !wDivisisbleByStride) {
+    return false;
+  }
+  auto isSymmetric = [](SmallVector<int64_t, 4> elements) -> bool {
+    return std::all_of(elements.begin(), elements.end(),
+        [&elements](int64_t i) { return (i == elements[0]); });
+  };
+  // one Phase Decomposition
+  if (stridesShape[0] == 1) {
+    return true;
+  }
+
+  bool fourPhaseDecomposition = (stridesShape[0] == 2);
+  bool ninePhaseDecomposition = (stridesShape[0] == 3);
+  if (fourPhaseDecomposition) {
+    if (kernelShape[0] == 6 && padsShape[0] == 2 && isSymmetric(padsShape)) {
+      // Currently support only with pads [2, 2, 2, 2]
+      return true;
+    }
+    if (kernelShape[0] == 3) {
+      // Supports [0,0,1,1] , [1,1,0,0] padding only.
+      if ((padsShape[0] == 0 && padsShape[1] == 0 && padsShape[2] == 1 &&
+              padsShape[3] == 1) ||
+          (padsShape[0] == 1 && padsShape[1] == 1 && padsShape[2] == 0 &&
+              padsShape[3] == 0)) {
+        return true;
+      }
+    }
+  }
+  if (ninePhaseDecomposition) {
+    // Supports only with padding [0, 0, 0, 0]
+    if (kernelShape[0] == 3 && isSymmetric(padsShape) && padsShape[0] == 0) {
+      return true;
+    }
+  }
+  return false;
+}
+Value computeConvTranspose(PatternRewriter &rewriter, Location loc,
+    ONNXConvTransposeOp op, Value input, Value weights, Value bias,
+    ArrayAttr dilations, IntegerAttr group, ArrayAttr kernel_shape,
+    ArrayAttr pads, ArrayAttr strides) {
+
+  RankedTensorType weightsType =
+      mlir::cast<RankedTensorType>(weights.getType());
+  assert(weightsType.hasRank() && "Weight tensor must have rank.");
+  Type elementType = getElementType(op.getType());
+  auto getVectorFromArrayAttr =
```

Comment:
```suggestion
  auto getVectorFromArrayAttr =
```
Please don't duplicate this function

Reply:
Removed duplication